lib/plenario.ex

```diff
@@ -4,46 +4,51 @@ defmodule Plenario do
   will surface data sets via searches.
   """
 
+  import Ecto.Query
+
+  import Geo.PostGIS, only: [st_intersects: 2]
+
+  import Plenario.Queries.Utils, only: [tsrange_intersects: 2]
+
   alias Plenario.Repo
 
   alias Plenario.Schemas.Meta
 
-  alias Plenario.Queries.MetaQueries
-
   @doc """
   Searches Metas whose bounding box intersects with the bounding box
   param, and optionally whose time range intersects with the time range param.
   """
-  @spec search_data_sets(bbox :: Geo.Polygon, time_range :: Postgrex.Range | nil) :: list(Meta)
-  def search_data_sets(bbox, time_range \\ nil) do
-    query =
-      case time_range == nil do
-        true ->
-          MetaQueries.list()
-          |> MetaQueries.handle_opts(
-              with_user: true,
-              with_charts: true,
-              ready_only: true,
-              bbox_intersects: bbox)
-
-        false ->
-          range =
-            case time_range do
-              %Plenario.TsRange{} ->
-                {:ok, range} = Plenario.TsRange.dump(time_range)
-                range
-              %Postgrex.Range{} ->
-                time_range
-            end
-          MetaQueries.list()
-          |> MetaQueries.handle_opts(
-              with_user: true,
-              with_charts: true,
-              ready_only: true,
-              bbox_intersects: bbox,
-              time_range_intersects: range)
-      end
-
-    Repo.all(query)
+  @spec search_data_sets(user_geom :: Geo.Polygon, time_range :: Postgrex.Range | nil) :: list(Meta)
+  def search_data_sets(user_geom, user_time_range \\ nil) do
+    search_bbox(user_geom, user_time_range)
+    |> Repo.all()
+    |> refine_to_hull(user_geom)
+    |> Repo.all()
+  end
```

Comment:
Not a necessary change, but I wonder if calculating intersections between complex geometries is expensive enough to justify prepping things with a bbox search. I like the idea of a zippy bbox search first, then a precise hull search second, but I wish I could find some _numbas_

Reply:
I don't know what the real world savings are for the dataset sizes we're working with, but comparing bboxes is O(1) while comparing hulls will be O(n*m) in the worst-case, where *n* and *m* are the number of vertices of the first and second polygon respectively.